// @babylonjs/serializers/glTF/2.0/Extensions/KHR_materials_clearcoat_color.js
import { GLTFExporter } from "../glTFExporter.js";
import { PBRBaseMaterial } from "@babylonjs/core/Materials/PBR/pbrBaseMaterial.js";
import { OpenPBRMaterial } from "@babylonjs/core/Materials/PBR/openpbrMaterial.js";
const NAME = "KHR_materials_clearcoat_color";
/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
export class KHR_materials_clearcoat_color {
    constructor(exporter) {
        /** Name of this extension */
        this.name = NAME;
        /** Defines whether this extension is enabled */
        this.enabled = true;
        /** Defines whether this extension is required */
        this.required = false;
        this._wasUsed = false;
        this._exporter = exporter;
    }
    dispose() { }
    /** @internal */
    get wasUsed() {
        return this._wasUsed;
    }
    async postExportMaterialAdditionalTexturesAsync(context, node, babylonMaterial) {
        const additionalTextures = [];
        if (babylonMaterial instanceof PBRBaseMaterial) {
            if (babylonMaterial.clearCoat.isEnabled) {
                if (babylonMaterial.clearCoat.tintTexture) {
                    additionalTextures.push(babylonMaterial.clearCoat.tintTexture);
                }
                return additionalTextures;
            }
        }
        else if (babylonMaterial instanceof OpenPBRMaterial) {
            if (babylonMaterial.coatWeight > 0) {
                if (babylonMaterial.coatColorTexture) {
                    additionalTextures.push(babylonMaterial.coatColorTexture);
                }
                return additionalTextures;
            }
        }
        return [];
    }
    // eslint-disable-next-line no-restricted-syntax
    postExportMaterialAsync(context, node, babylonMaterial) {
        return new Promise((resolve) => {
            if (babylonMaterial instanceof PBRBaseMaterial) {
                if (!babylonMaterial.clearCoat.isEnabled) {
                    resolve(node);
                    return;
                }
                this._wasUsed = true;
                // This material must have the clearcoat extension already before
                // we can add the clearcoat color sub-extension
                const parentExt = node.extensions ? node.extensions["KHR_materials_clearcoat"] : null;
                if (!parentExt) {
                    resolve(node);
                    return;
                }
                const coatColorTextureInfo = this._exporter._materialExporter.getTextureInfo(babylonMaterial.clearCoat.tintTexture);
                const clearCoatInfo = {
                    clearcoatColorFactor: babylonMaterial.clearCoat.tintColor.asArray(),
                    clearcoatColorTexture: coatColorTextureInfo ?? undefined,
                };
                if (clearCoatInfo.clearcoatColorTexture !== null) {
                    this._exporter._materialNeedsUVsSet.add(babylonMaterial);
                }
                parentExt.extensions = parentExt.extensions || {};
                parentExt.extensions[NAME] = clearCoatInfo;
            }
            else if (babylonMaterial instanceof OpenPBRMaterial) {
                if (babylonMaterial.coatWeight == 0.0) {
                    resolve(node);
                    return;
                }
                this._wasUsed = true;
                // This material must have the clearcoat extension already before
                // we can add the clearcoat color sub-extension
                const parentExt = node.extensions ? node.extensions["KHR_materials_clearcoat"] : null;
                if (!parentExt) {
                    resolve(node);
                    return;
                }
                const coatColorTextureInfo = this._exporter._materialExporter.getTextureInfo(babylonMaterial.coatWeightTexture);
                const clearCoatInfo = {
                    clearcoatColorFactor: babylonMaterial.coatColor.asArray(),
                    clearcoatColorTexture: coatColorTextureInfo ?? undefined,
                };
                if (clearCoatInfo.clearcoatColorTexture !== null) {
                    this._exporter._materialNeedsUVsSet.add(babylonMaterial);
                }
                parentExt.extensions = parentExt.extensions || {};
                parentExt.extensions[NAME] = clearCoatInfo;
            }
            resolve(node);
        });
    }
}
GLTFExporter.RegisterExtension(NAME, (exporter) => new KHR_materials_clearcoat_color(exporter), 105);
